File: code_snippets/python/src/develop/retries.py
from datetime import timedelta

import restate
from restate import Context, Service
from restate.exceptions import TerminalError

my_service = Service("MyService")


def write_to_other_system():
    pass


@my_service.handler("myHandler")
async def my_service_handler(ctx: Context, greeting: str) -> str:
    # <start_here>
    try:
        await ctx.run(
            "write",
            lambda: write_to_other_system(),
            # <mark_1>
            # Max number of retry attempts to complete the action.
            max_attempts=3,
            # Max duration for retrying, across all retries.
            max_retry_duration=timedelta(seconds=10),
            # </mark_1>
        )
    except TerminalError as err:
        # Handle the terminal error after retries exhausted
        # For example, undo previous actions (see sagas guide) and
        # propagate the error back to the caller
        raise err
    # <end_here>

    return f"${greeting}!"


app = restate.app([my_service])

File: code_snippets/ts/src/develop/retries.ts
import * as restate from "@restatedev/restate-sdk";
import { RestatePromise } from "@restatedev/restate-sdk";
import { MyService } from "./my_service";
import { RunOptions } from "@restatedev/restate-sdk/dist/esm/src/context";

function writeToOtherSystem() {
  return undefined;
}

const service = restate.service({
  name: "MyService",
  handlers: {
    greet: async (ctx: restate.Context, name: string) => {
      // <start_here>
      try {
        // <mark_1>
        const myRunRetryPolicy = {
          initialRetryIntervalMillis: 500,
          retryIntervalFactor: 2,
          maxRetryIntervalMillis: 1000,
          maxRetryAttempts: 5,
          maxRetryDurationMillis: 1000,
        };
        // </mark_1>
        await ctx.run("write", () => writeToOtherSystem(), myRunRetryPolicy);
      } catch (e) {
        if (e instanceof restate.TerminalError) {
          // Handle the terminal error after retries exhausted
          // For example, undo previous actions (see sagas guide) and
          // propagate the error back to the caller
        }
        throw e;
      }
      // <end_here>
    },
  },
});

const paymentClient = {
  call: (txId: string, amt: number) => {
    return { error: "", isSuccess: false };
  },
};
